File: subscribe.js
const mqtt = require('mqtt');
const { Client } = require('pg');
require('dotenv').config();

const mqttOptions = {
    host: process.env.MQTT_HOST,
    port: process.env.MQTT_PORT,
    protocol: 'mqtts',
    username: process.env.MQTT_USERNAME,
    password: process.env.MQTT_PASSWORD
};

const pgClient = new Client({
    connectionString: process.env.DATABASE_URL,
    ssl: { rejectUnauthorized: false }  
});

pgClient.connect()
    .then(() => console.log('Connected to NeonSQL'))
    .catch(err => console.error('NeonSQL Connection Error:', err));

const mqttClient = mqtt.connect(mqttOptions);

mqttClient.on('connect', function () {
    console.log('Connected to MQTT');
    mqttClient.subscribe(process.env.MQTT_TOPIC, function (err) {
        if (!err) {
            console.log('Subscribed to SENSOR/DATA topic');
        }
    });
});

mqttClient.on('error', function (error) {
    console.error('MQTT Error:', error);
});

async function insertIntoSQL(data) {
    try {
        const query = `
            INSERT INTO water_quality (
                area, areaname, block, ph, turbidity, watertemperature, 
                totaldissolvedsolids, conductivity, chlorine, fluoride, target
            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
        `;

        const values = [
            data.area, 
            data.areaname, 
            data.block, 
            parseFloat(data.ph), 
            parseFloat(data.turbidity), 
            parseFloat(data.watertemperature), 
            parseFloat(data.totaldissolvedsolids), 
            parseFloat(data.conductivity), 
            parseFloat(data.chlorine), 
            parseFloat(data.fluoride), 
            parseFloat(data.target)
        ];

        await pgClient.query(query, values);
        console.log("Data inserted into NeonSQL successfully.");
    } catch (error) {
        console.error("NeonSQL Insertion Error:", error);
    }
}


mqttClient.on('message', function (topic, message) {
    console.log('Received message:', topic, message.toString());

    try {
        const jsonString = message.toString().replace(/'/g, '"'); 
        const data = JSON.parse(jsonString);

        console.log("Parsed Data:", data);

        insertIntoSQL(data);
    } catch (error) {
        console.error('Message Parsing Error:', error);
    }
});

File: subscribe.py
import json
import ssl
import paho.mqtt.client as mqtt
import psycopg2
import os
from dotenv import load_dotenv
from data_validation import valid_data
from predict import predict

load_dotenv()  

pg_conn = psycopg2.connect(os.getenv("DATABASE_URL"), sslmode='require')
pg_cursor = pg_conn.cursor()

mqtt_broker = os.getenv("MQTT_HOST")
mqtt_port = int(os.getenv("MQTT_PORT", 8883))
mqtt_username = os.getenv("MQTT_USERNAME")
mqtt_password = os.getenv("MQTT_PASSWORD")
mqtt_topic = os.getenv("MQTT_TOPIC")

def on_connect(client, userdata, flags, rc):
    if rc == 0:
        print("✅ Connected to MQTT Broker")
        client.subscribe(mqtt_topic)
        print(f"📡 Subscribed to topic: {mqtt_topic}")
    else:
        print(f"❌ Connection failed with code {rc}")

def on_message(client, userdata, message):
    json_string = message.payload.decode().replace("'", '"')  
    data = json.loads(json_string)
    if valid_data(data):
        y_pred = predict(data)
        data['target'] = y_pred
        insert_into_db(data)
        print("✅ Parsed Data:", data)
    

def insert_into_db(data):
    try:
        query = """
            INSERT INTO water_quality (
                area, areaname, block, ph, turbidity, watertemperature, 
                totaldissolvedsolids, conductivity, chlorine, fluoride, target
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        """
        values = (
            data["area"], 
            data["areaname"], 
            data["block"], 
            float(data["ph"]), 
            float(data["turbidity"]), 
            float(data["watertemperature"]), 
            float(data["totaldissolvedsolids"]), 
            float(data["conductivity"]), 
            float(data["chlorine"]), 
            float(data["fluoride"]), 
            float(data["target"])
        )

        print("📊 Preparing to insert data into the database...")
        print(f"📌 Insert Values: {values}")

        pg_cursor.execute(query, values)
        pg_conn.commit()
        print("✅ Data successfully inserted into NeonSQL.")

    except Exception as e:
        print(f"❌ Database Insertion Error: {e}")

mqtt_client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION1)
mqtt_client.tls_set_context(ssl.SSLContext(ssl.PROTOCOL_TLS))
mqtt_client.username_pw_set(mqtt_username, mqtt_password)
mqtt_client.on_connect = on_connect
mqtt_client.on_message = on_message

print("🚀 Connecting to MQTT broker...")
mqtt_client.connect(mqtt_broker, mqtt_port)

mqtt_client.loop_forever()
